### PR TITLE
Fix races in AppInitMain and others with lock and atomic bools

### LINKED ISSUE #11106: Race in printing mapBlockIndex.size() in init
init.cpp:1635 prints mapBlockIndex.size() without cs_main. This is technically a race (though it is rather inconsequential). At some point someone should fix this so tsan stops complaining.

### PR DESCRIPTION
**Pull request by meshcollider**
Fixes https://github.com/bitcoin/bitcoin/issues/11106

Also makes fReindex atomic as suggested in @TheBlueMatt comment below, and makes fUseCrypto atomic as suggested in 10916

https://github.com/bitcoin/bitcoin/pull/11107/commits/d291e7635b0ef4156c2805c6c4ee1adad91f0307 just renames the parameters in the txdb header file to make them consistent with those used in the cpp file, noticed it when looking for uses of fReindex

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -1632,8 +1632,12 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
     // ********************************************************* Step 11: start node
 
     //// debug print
-    LogPrintf("mapBlockIndex.size() = %u\n",   mapBlockIndex.size());
-    LogPrintf("nBestHeight = %d\n",                   chainActive.Height());
+    {
```
Comment (jonasschnelli) [inline src/init.cpp]: I guess it would make sense to also cover `mapBlockIndex` and `chainActive` to make this move-safe.

Comment (MeshCollider) [inline src/init.cpp]: chainActive.height() is also accessed on line 1652, you want to bring it in too?

**Comment by MeshCollider** [comment]
Travis build looks unrelated, timeout in sendheaders.py - `assert(wait_until(test_function, timeout=60))`

File: src/init.cpp (hunk 1 above)
Comment (promag) [inline src/init.cpp]: If chainActive is added then update descriptions accordingly.

Did you search for other cases?

Comment (MeshCollider) [inline src/init.cpp]: No haven't modified the pull yet, it still just fixes the issue linked

**Comment by TheBlueMatt** [comment]
While you're at it, want to make fReindex atomic so that the ThreadImport write doesn't "race" with net_processing reads?

**Comment by MeshCollider** [comment]
@TheBlueMatt want to review?

**Comment by MeshCollider** [comment]
Ping @TheBlueMatt if you have a bit of time :)

File: src/init.cpp [hunk 2]
```diff
@@ -1632,8 +1632,12 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
...
+        LOCK(cs_main);
+        LogPrintf("mapBlockIndex.size() = %u\n", mapBlockIndex.size());
+    }
+    LogPrintf("nBestHeight = %d\n", chainActive.Height());
+
```
Comment (TheBlueMatt) [inline src/init.cpp]: nit: EOL whitespace here.

File: src/validation.cpp [hunk 3]
```diff
@@ -3523,7 +3523,7 @@ bool static LoadBlockIndexDB(const CChainParams& chainparams)
     // Check whether we need to continue reindexing
     bool fReindexing = false;
     pblocktree->ReadReindexing(fReindexing);
-    fReindex |= fReindexing;
+    fReindex = fReindex | fReindexing;
```
Comment (TheBlueMatt) [inline src/validation.cpp]: Might as well use the atomic or here since we're making it atomic, no?

File: src/init.cpp (hunk 1 above)
Comment (TheBlueMatt) [reply src/init.cpp]: Indeed, ight as well pull chainActive.Height() into the cs_main here.

File: src/validation.cpp (hunk 3 above)
Comment (MeshCollider) [reply src/validation.cpp]: `|=`? It is only defined on atomic integral types I believe, not bool

Comment (TheBlueMatt) [reply src/validation.cpp]: Err, indeed, quivalent is to do an if(fReindexing) fReindex = true

Comment (MeshCollider) [reply src/validation.cpp]: Yeah true that's cleaner, will do

**Comment by TheBlueMatt** [comment]
utACK 731065b114452ff770320d09639448b3c9a74b0a

File: src/validation.cpp [hunk 4]
```diff
@@ -3523,7 +3523,7 @@ bool static LoadBlockIndexDB(const CChainParams& chainparams)
     // Check whether we need to continue reindexing
     bool fReindexing = false;
     pblocktree->ReadReindexing(fReindexing);
-    fReindex |= fReindexing;
```
Comment (jtimon) [inline src/validation.cpp]: What's wrong with |= ?

**Comment by jtimon** [comment]
utACK 731065b

File: src/validation.cpp (hunk 4 above)
Comment (MeshCollider) [reply src/validation.cpp]: See my reply to @TheBlueMatt's comment above, it's not implemented for atomic bool :)

**Comment by MeshCollider** [comment]
Added a new commit to make fUseCrypto atomic as well, see https://github.com/bitcoin/bitcoin/pull/10916#issuecomment-317340990

**Comment by MeshCollider** [comment]
Friendly ping @jonasschnelli and/or @TheBlueMatt if you have time, the last commit is only a 2 line change so shouldn't be too time consuming to review ;)

**Comment by TheBlueMatt** [comment]
utACK c626dcb50eed496462fd4ac3e05bf79164749ebe

**Comment by jonasschnelli** [comment]
utACK https://github.com/bitcoin/bitcoin/commit/c626dcb50eed496462fd4ac3e05bf79164749ebe
Generally "unrelated" cleanups (like 731065b114452ff770320d09639448b3c9a74b0a) should be avoided.

**Pull request merged**